### DISCUSSION TITLE
Tomcat logging and geoserver monitoring/audit questions

### DISCUSSION THREAD

**Discussion by alexkuretz** [kartoza/docker-geoserver#624] [Q&A] [answered]
Hi, we've been using your geoserver image for a long time, thank you for your work on this! I've recently put the newer 2.24.0 release into our production environment and am noticing some issues (for us) with the way tomcat logging is configured, as well as with the audit/monitor functionality of geoserver. 

We are generating >1GB/day of tomcat localhost_access* logs, the default logger is set to hold 90 days worth. I don't see a way to manage this configuration as part of the kartoza image, is there an option I'm missing? I had an idea to add `java.util.logging.config.file` system property to the `ADDITIONAL_JAVA_STARTUP_OPTIONS` pointing it to my modified logging.properties file but I haven't tried it yet.

Also, we had a production outage due to geoserver taking a very long time to start, it  turns out the audit functionality had generated tens of thousands of audit log files, and [as part of the startup it deletes everything](https://github.com/kartoza/docker-geoserver/blob/c65bda32f59601f1ea2384f64536fe4ab23c5796/scripts/start.sh#L162) in the $DATADIR/monitoring/monitor_* log directory. I think the [default monitoring.properties file configuration](https://github.com/kartoza/docker-geoserver/blob/c65bda32f59601f1ea2384f64536fe4ab23c5796/scripts/functions.sh#L555) isn't ideal for production use, and my changes to the file get overwritten when the geoserver pod restarts. It looks like if I deliver my own monitor.properties file in /settings I can control the behavior, but am curious why the default is set this way, with rolling the file every 40 requests? 

FInally, unless I'm mistaken, it appears that the geoserver logging [will get reset on start](https://github.com/kartoza/docker-geoserver/blob/c65bda32f59601f1ea2384f64536fe4ab23c5796/scripts/functions.sh#L396) every time, unless I specify an ([undocumented](https://github.com/kartoza/docker-geoserver/blob/master/.env)) env variable `GEOSERVER_LOG_LEVEL`. Is this intentional?

Thanks again, and I look forward to your thoughts on these items.

**Comment by NyakudyaA** [comment, marked as answer]
A couple of answers here
* Default monitoring properties isn't ideal for production - We rely on power users like you to annotate what you think are the desirable defaults. What can the default be and how can we set this? Alternatively, you can add a couple of env variables like
```
audit.enabled=${MONITORING_AUDIT_ENABLED}
audit.roll_limit=${MONITORING_AUDIT_ROLL_LIMIT}
```
* Large tomcat logs - You can configure docker-compose like the following
```
logging:
      driver: json-file
      options:
        max-size: 200m
        max-file: '10'
```
The startup options can also be set but I am not sure what options you will need to add there.
* Audit functionality deletes everything on startup - We can safely remove this option. The assumption here was it's necessary to cleanup the logs. We should ideally set this as an option and in production env should you not be logging against a production DB
```
if [[ -d "${GEOSERVER_DATA_DIR}"/monitoring ]];then
   if [ ${RESET_MONITORING_LOGS} == 'True' ]];then
    find "${GEOSERVER_DATA_DIR}"/monitoring -type d -name 'monitor_*' -exec rm -r {} +
   fi
  fi
```
If you think the answers above will help you can proceed to do a PR

**Comment by NyakudyaA** [comment]
@alexkuretz Please see https://github.com/kartoza/docker-geoserver/commit/3c01ba16cdf1d847db1a88eb1da565b5fe7ea25c that resolves the majority of your issues

**Comment by alexkuretz** [comment]
Thanks for that, I was about half-way there with a PR but hadn't finished up yet. The only immediate suggestion I'd have is to default to audit disabled, as that [is the default](https://docs.geoserver.org/main/en/user/extensions/monitoring/audit.html) in Geoserver. I'll give this a more thorough look and try it out.

> The audit.enable is used to turn on the logger (it is off by default).

**Comment by alexkuretz** [comment]
Also, in the PR I was working on I was adding these new options to the .env file so they'd be exposed to users, so far I'd added this at line 84:
```
# Geoserver monitoring and audit logging https://docs.geoserver.org/main/en/user/extensions/monitoring/audit.html
MONITORING_AUDIT_ENABLED=false
MONITORING_AUDIT_ROLL_LIMIT=20
MONITORING_RESET_LOGS=false
```

I'm happy to continue working on my PR if that would be helpful, do you prefer I make comment suggestions to your PR or just submit my own in the next day or two?

**Reply by NyakudyaA** [reply]
You can continue with your PR, just make sure you pull from develop first and then do the PR

**Comment by alexkuretz** [comment]
I've submitted the PR and look forward to your feedback. https://github.com/kartoza/docker-geoserver/pull/627

**Comment by alexkuretz** [comment]
We had another production outage this week when the server paused for a long time deleting 200k+ audit files after a reboot, really need to get the fix for overwriting the monitor.properties file that's in my PR, any idea when it might get reviewed/accepted?

**Reply by NyakudyaA** [reply]
@alexkuretz what image are using,since I merged that PR if should be not deleting. I also reviewed your PR and now waiting for your feedback

**Reply by alexkuretz** [reply]
We build our images off of your base image, currently on 2.24.0 for production and waiting on the fixes in my PR to roll a new production based off of your latest 2.24.2 image. It's of course complicated by the fact that [your tagged images aren't stable](https://github.com/kartoza/docker-geoserver/issues/233) . Also, current `develop` branch [defaults to enabling the audit](https://github.com/kartoza/docker-geoserver/blob/3c01ba16cdf1d847db1a88eb1da565b5fe7ea25c/scripts/env-data.sh#L420) and [overwrites the config file](https://github.com/kartoza/docker-geoserver/blob/3c01ba16cdf1d847db1a88eb1da565b5fe7ea25c/scripts/functions.sh#L561) on startup, which is not the geoserver default and will fill up the disk with hundreds of thousands of tiny files, so even though the default is not to delay startup by deleting them, it still is not a good solution. My PR fixes these issues. 

For clarity, we run in kubernetes so if the pod restarts the startup scripts all run and reset things in undesirable ways, which is the purpose of my PR. I don't think existing config files should be blindly overwritten by startup scripts.

I never got any notifications and don't see any review comments on my PR? Happy to provide feedback as soon as I see a question!
<img width="950" alt="image" src="https://github.com/kartoza/docker-geoserver/assets/100612600/edbe1c1a-3843-48b5-8a5c-bdc3806ca712">

**Reply by NyakudyaA** [reply]
@alexkuretz The images are stable now. You will see that using a version like 2.24.2, when the action is pushed it builds a new version but if you choose to use an image with a date i.e 2.24.2-v23-04 etc. That specific image will never get replaced again by the action.

Please check your PR there is a comment I left there. Let me try to ping you

**Reply by alexkuretz** [reply]
Ok, thanks for that, I'll look at using the dated images instead of versioned. 

I don't see any comments on the PR.
<img width="565" alt="image" src="https://github.com/kartoza/docker-geoserver/assets/100612600/e5c5f675-3a18-4f21-a839-387ee2fbc90a">
